spec/models/cookie_student_spec.rb

```diff
@@ -0,0 +1,196 @@
+require 'rails_helper'
+
+RSpec.describe CookieStudent, type: :model do
+  def approvable_ids_in_cookie(cookie)
+    value = cookie[:approved_approvable_ids]
+    value.present? ? JSON.parse(value) : nil
+  end
+
+  describe '#add' do
+    it 'adds approvable.id only if available' do
+      subject1 = create :subject, :with_exam
+      subject2 = create :subject, :with_exam
+
+      create(:subject_prerequisite, approvable: subject2.course, approvable_needed: subject1.course)
+
+      cookies = build(:cookie)
+      student = build(:cookie_student, cookies:)
+
+      student.add(subject2.course)
+      expect(approvable_ids_in_cookie(cookies)).to be_nil
+
+      student.add(subject1.course)
+      expect(approvable_ids_in_cookie(cookies)).to eq([subject1.course.id])
+
+      student.add(subject2.course)
+      expect(approvable_ids_in_cookie(cookies)).to eq([subject1.course.id, subject2.course.id])
+    end
+
+    it 'for a subject.exam adds subject.course as well' do
+      subject = create :subject, :with_exam
+      cookies = build(:cookie)
+      student = build(:cookie_student, cookies:)
+      student.add(subject.exam)
+
+      expect(approvable_ids_in_cookie(cookies)).to eq([subject.exam.id, subject.course.id])
+    end
+  end
+
+  describe '#remove' do
+    it 'removes approvable.id and just the exam of the subject' do
+      subject1 = create :subject, :with_exam
+      subject2 = create :subject, :with_exam
+      subject3 = create :subject, :with_exam
+      subject4 = create :subject, :with_exam
+
+      create(:subject_prerequisite, approvable: subject2.course, approvable_needed: subject3.course)
+      create(:subject_prerequisite, approvable: subject3.course, approvable_needed: subject1.course)
+
+      cookies = build(:cookie,
+                      approved_approvable_ids: [
+                        subject1.course.id,
+                        subject2.course.id,
+                        subject3.course.id,
+                        subject4.course.id
+                      ])
+      student = build(:cookie_student, cookies:)
+      student.remove(subject1.course)
+
+      expect(approvable_ids_in_cookie(cookies)).to eq([subject2.course.id, subject3.course.id, subject4.course.id])
+    end
+  end
+
+  describe '#available?' do
+    it 'returns true if subject_or_approvable is available' do
+      subject1 = create :subject, :with_exam
+      create(:subject_prerequisite, approvable: subject1.exam, approvable_needed: subject1.course)
+
+      expect(build(:cookie_student, approved_approvable_ids: []).available?(subject1)).to be true
+      expect(build(:cookie_student, approved_approvable_ids: []).available?(subject1.course)).to be true
+      expect(build(:cookie_student, approved_approvable_ids: []).available?(subject1.exam)).to be false
+
+      cookies = build(:cookie, approved_approvable_ids: [subject1.course.id])
+      expect(build(:cookie_student, cookies:).available?(subject1)).to be true
```

Comment:
Debería ser `expect(build(:cookie_student, cookies:).available?(subject1.exam)).to be true`?